src/entities/abTrail/ui/AbCanvas.tsx

```diff
@@ -252,63 +219,59 @@ const AbCanvas: FC<Props> = props => {
     return greenPointIndex;
   };
 
-  const onTouchStart = (touchInfo: TouchInfo) => {
+  const onTouchStart = (x: number, y: number, time: number) => {
     const isFinished = paths.length === currentIndexRef.current;
 
     if (currentPathRef.current || readonly || isFinished) {
+      resetCurrentPath();
       return;
     }
 
-    const point: Point = { x: touchInfo.x, y: touchInfo.y };
+    const point: Point = { x, y };
 
     if (isOverAny(point) && !isOverCurrent(point)) {
       onMessage(MessageType.IncorrectStartPoint);
+      resetCurrentPath();
       return;
     }
 
     if (!isOverCurrent(point)) {
+      resetCurrentPath();
       return;
     }
 
     addLogLine(point);
     reCreatePath(point);
-    drawPath();
-    reRender();
+
     onLog({
-      x: (touchInfo.x * width) / 100,
-      y: (touchInfo.y * width) / 100,
-      time: Date.now(),
+      x: (x * width) / 100,
+      y: (y * width) / 100,
+      time: time,
     });
   };
 
-  const onTouchProgress = (touchInfo: TouchInfo) => {
+  const onTouchProgress = (x: number, y: number, time: number) => {
     const currentPath = getCurrentPath();
 
+    if (errorPath) {
+      resetCurrentPath();
+      return;
+    }
+
     if (!currentPath) {
       return;
     }
 
-    const point: Point = { x: touchInfo.x, y: touchInfo.y };
+    const point: Point = { x, y };
 
     currentPath.lineTo(point.x, point.y);
 
     addLogPoint(createLogPoint(point));
 
-    drawPath();
-
-    if (isCloseToNext(point) && !isCloseToNextRerendered()) {
-      reRender();
-      setCloseToNextRerendered();
-    }
-
-    if (!isCloseToNext(point)) {
-      resetCloseToNextRerendered();
-    }
-
     onLog({
```

Comment:
Please, add local function: addPointToStream and move call of onLog in there for all onLog calls.

Reply:
Done